pygam/pygam.py

```diff
@@ -1870,6 +1876,260 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
+              input data
+
+        y : array of shape (n_samples,)
+              response vector
+
+        weights : np.array of shape (n_samples,)
+            sample weights
+
+        n_draws : positive int, default: 100
+            The number of samples to draw from the posterior distribution of
+            the coefficients and smoothing parameters
+
+        n_bootstraps : positive int, default: 1
+            The number of bootstrap samples to draw from simulations of the
+            response (from the already fitted model) to estimate the
+            distribution of the smoothing parameters given the response data.
+            If `n_bootstraps` is 1, then only the already fitted model's
+            smoothing parameters is used.
+
+        objective : string, default: 'auto'
+            metric to optimize in grid search. must be in
+            ['AIC', 'AICc', 'GCV', 'UBRE', 'auto']
+            if 'auto', then grid search will optimize GCV for models with
+            unknown scale and UBRE for models with known scale.
+
+        Returns
+        -------
+        coef_samples : array of shape (n_draws, n_samples)
+            Approximate simulations of the coefficients drawn from the
+            posterior distribution of the coefficients and smoothing
+            parameters given the response data
+
+        References
+        ----------
+        Simon N. Wood, 2006. Generalized Additive Models: an introduction with
+        R. Section 4.9.3 (pages 198–199) and Section 5.4.2 (page 256–257).
+        """
+        if not self._is_fitted:
+            raise AttributeError('GAM has not been fitted. Call fit first.')
+        if n_bootstraps < 1:
+            raise ValueError('n_bootstraps must be >= 1;'
+                             ' got {}'.format(n_bootstraps))
+        if n_draws < 1:
+            raise ValueError('n_draws must be >= 1;'
+                             ' got {}'.format(n_draws))
+
+        coef_bootstraps, cov_bootstraps = (
+            self._bootstrap_samples_of_smoothing(X, y, weights=weights,
+                                                 n_bootstraps=n_bootstraps,
+                                                 objective=objective))
+        coef_draws = self._simulate_coef_from_bootstraps(
+            n_draws, coef_bootstraps, cov_bootstraps)
+
+        return coef_draws
+
+    def _bootstrap_samples_of_smoothing(self, X, y, weights=None,
+                                        n_bootstraps=1, objective='auto'):
+        """Sample the smoothing parameters using simulated response data."""
+        mu = self.predict_mu(X)  # Wood pg. 198 step 1
+        coef_bootstraps = [self.coef_]
+        cov_bootstraps = [
+            _make_positive_semi_definite(self.statistics_['cov'])]
+
+        for _ in range(n_bootstraps - 1):  # Wood pg. 198 step 2
+            # generate response data from fitted model (Wood pg. 198 step 3)
+            y_bootstrap = self.distribution.sample(mu)
```

Comment:
hmm. actually im wondering about a technicality:
- wood says to sample errors/deviates from the fitted distribution
- add those errors to the original responses
- these are now our bootstraps

it seems the approach in the code is slightly different.
- we are essentially sampling deviates and adding them to mu instead of y

i think we will probably get more variation if we add deviates to the original responses?
what do you think?

Reply:
(reading pg 199, sec 4.9.3, step 3)

Reply:
you know what, never mind.

i think what you have is correct

Reply:
```
Based on the fitted model from step 1, simulate a parametric bootstrap response vector y[k]. That is, simulate random deviates with the appropriate response distribution, and mean given by the fitted values from step 1.
```

sounds like this means to sample from the original model. ie the way you have it.